drivers: timer: nrf: Optimize k_cycle_get_32().

pizi-nordic opened the pull request:
The previous implementation of _timer_cycle_get_32() (which is directly
mapped to k_cycle_get_32()) taken from 87 to 132 cycles. As result it
was too heavy for using it as source of time for logger.

This commit makes this function faster by removing redundant access
to the RTC register (each access consumed 28 CPU cycles) as well as
loop, which made this call non-deterministic.

After these changes the k_cycle_get_32() needs only 50-52 cycles
in to calculate 32-bit timestamp.

This PR fixes #8798.

pizi-nordic committed:
drivers: timer: nrf: Optimize k_cycle_get_32().

The previous implementation of _timer_cycle_get_32() (which is directly
mapped to k_cycle_get_32()) taken from 87 to 132 cycles. As result it
was too heavy for using it as source of time for logger.

This commit makes this function faster by removing redundant access
to the RTC register (each access consumed 28 CPU cycles) as well as
loop, which made this call non-deterministic.

After these changes the k_cycle_get_32() needs only 50-52 cycles
in to calculate 32-bit timestamp.

Signed-off-by: Piotr Zięcik <piotr.ziecik@nordicsemi.no>

carlescufi commented:
recheck

pizi-nordic commented on drivers/timer/nrf_rtc_timer.c:
```diff
@@ -550,24 +550,33 @@ int _sys_clock_driver_init(struct device *device)
-				 RTC_MASK;
-		rtc_prev = rtc_now;
-		rtc_now = RTC_COUNTER;
-	} while (rtc_now != rtc_prev);
```
I don't think that such "protection" gives any value:

```
int get_time()
{
    int time;

/*
 * If RTC_REGISTER changes here, code below will get updated value.
 */
    do {
        time = RTC_REGISTER;
        /*
         *  If RTC_REGISTER changes here, the loop will actually delay return
         * from the function and returned value will include effect of the delay.
         */
     } while (time != RTC_REGISTER)

/*
 * If RTC_REGISTER changes here, value returned by this function will be off by 1 system clock tick.
 */

return time;
}
```

In both versions, the one with loop and the one without there is risk that a "stale" value will be returned. It is unavoidable, as even return from the function takes time and RTC counter could be updated during this time. I personally do not see any benefit from this loop (especially as in this case access to RTC is atomic and there is no risk that we get inconsistent value of the register).

cvinayak replied:
What you are not considering is the one clock tick that elapses, and the kernel cannot tolerate a stale clock tick. You can use your implementation on nRF52 boards and run a typical peripheral sample, it will stall eventually when the system clock ticks between the reading of the kernel variable and reading of the RTC counter.

pizi-nordic replied:
Are you mentioning here problem you pointed there: https://github.com/zephyrproject-rtos/zephyr/pull/9275#discussion_r207704157 ? If so, please look on my comment posted here: https://github.com/zephyrproject-rtos/zephyr/pull/9275#discussion_r207866959

cvinayak replied:
I understand your explanation. except for:
/*
 * If RTC_REGISTER changes here, value returned by this function will
 *  be off by 1 system clock tick.
 */

Here the returned value is only off by 1 cycle, not 1 system clock tick.

But as long as, tests cases pass as before, I am ok with the changes (because I am not very well informed on the kernel's use of this function).

pizi-nordic commented on drivers/timer/nrf_rtc_timer.c:
```diff
@@ -550,24 +550,33 @@ int _sys_clock_driver_init(struct device *device)
+	 * outdated at this point. However the value of rtc_count will
+	 * include missed ticks.
+	 */
+	rtc_count = RTC_COUNTER;
```
I think that it is compensated by rtc_count. Please note that the `_sys_clock_tick_count` is read before the `RTC_COUNTER` register (this order is enforced by compiler barrier). As `RTC_COUNTER` is _not_ reset on tick, it will include the missed time.

Please look at the following example:
````
RTC_COUNTER value:                        100           200           300
tick:                                                       1               2               3
_sys_clock_tick_count read:                             x
RTC_COUNTER read:                                                       x
````

In this example we we read `_sys_clock_tick_count` somewhere between tick 1 and 2. The read value was 1. Then we got an interrupt, which delayed `RTC_COUNTER` read (which returned 250, as RTC is counting all the time). The final calculation uses cached value of `_sys_clock_tick_count` (which is stale in this example) and current value of `RTC_COUNTER`:

ticked_cycles = cached _sys_clock_tick_count * RTC_COUNTS_PER_TICK = 1 * 100 = 100
elapsed_cycles = RTC_COUNT - ticked_cycles = 250 - 100 = 150
(please note that elapsed_cycles is larger than one tick).

total_time = 100 + 150 = 250  (correct)

pizi-nordic commented on drivers/timer/nrf_rtc_timer.c:
```diff
@@ -550,24 +550,33 @@ int _sys_clock_driver_init(struct device *device)
+	ticked_cycles = sys_clock_tick_count * sys_clock_hw_cycles_per_tick;
+
+	/* Number of timer cycles since last announced tick we know about. */
+	elapsed_cycles = rtc_count - (ticked_cycles & RTC_MASK);
```
Fixed. Thank you.
Feel free to look at the updated code.

pizi-nordic commented on drivers/timer/nrf_rtc_timer.c:
```diff
@@ -550,24 +550,33 @@ int _sys_clock_driver_init(struct device *device)
-		rtc_now = RTC_COUNTER;
-	} while (rtc_now != rtc_prev);
+	/* Get number of announced ticks. */
+	sys_clock_tick_count = _sys_clock_tick_count;
```
I tried to explain the whole idea, however you are right that the code was a bit too long for the job.
Feel free to look at the updated code.

anangl commented on drivers/timer/nrf_rtc_timer.c:
```diff
@@ -550,24 +550,27 @@ int _sys_clock_driver_init(struct device *device)
+
+	/* Number of timer cycles since last announced tick we know about.
+	 *
+	 * The value of RTC_COUNTER is not reset on tick, so it could
```
Not only _it could_, but even _it will_ compensate this.

pizi-nordic replied:
Fixed

anangl commented on drivers/timer/nrf_rtc_timer.c:
```diff
@@ -550,24 +550,27 @@ int _sys_clock_driver_init(struct device *device)
+	 * which could have happen between the ticked_cycles calculation
+	 * and the code below.
+	 */
+	elapsed_cycles = (RTC_COUNTER - (ticked_cycles & RTC_MASK)) & RTC_MASK;
```
Do you really need to do this masking twice? I think that only the final one will be sufficient (the upper bits are discarded anyway, so it doesn't matter if they are all zeros, all ones, or a mix of these).

pizi-nordic replied:
@cvinayak: What is your opinion about this?

cvinayak replied:
I agree with @anangl, single overall masking is sufficient.

pizi-nordic replied:
I thought so, however I wanted to hear your opinion before I will do the change. Fixed.

cvinayak commented:
@pizi-nordic could you check #9310 , which is failing even before your PR over master, this was the test suite that used to fail prior to the changes done for #3665 in #374

pizi-nordic commented:
@cvinayak: https://github.com/zephyrproject-rtos/zephyr/pull/9137 should fix #9310.

cvinayak commented:
@pizi-nordic could you also change the implementation in lines 278, 336, and 379 in a separate commit.

pizi-nordic committed:
driver: timer: nrf: Remove redundant RTC_COUNTER reads.

This commit removes redundant reads of RTC_COUNTER register
propagating optimizations made in the k_cycle_get_32() function.

Signed-off-by: Piotr Zięcik <piotr.ziecik@nordicsemi.no>

pizi-nordic commented:
@cvinayak: Done, please check.

anangl commented on drivers/timer/nrf_rtc_timer.c:
```diff
@@ -550,24 +529,27 @@ int _sys_clock_driver_init(struct device *device)
-	return (sys_clock_tick_count * sys_clock_hw_cycles_per_tick) +
-	       elapsed_cycles;
+	/* Number of timer cycles announced as ticks so far. */
+	ticked_cycles = _sys_clock_tick_count * sys_clock_hw_cycles_per_tick;
```
Now I realized that this information was stored in `rtc_past`, so actually there's no need to do this multiplication here.

pizi-nordic replied:
There is subtle difference:
```
rtc_past = (_sys_clock_tick_count * sys_clock_hw_cycles_per_tick) & RTC_MASK;
ticked_cycles = (_sys_clock_tick_count * sys_clock_hw_cycles_per_tick);
```
We are doing this multiplication because we want 32-bit timestamp which will not wrap up when RTC do so.

anangl replied:
However, ;-)
it seems that `rtc_past` could be stored without prior masking (then it could be used instead of the multiplication mentioned above), as in all places where it is later used the masking would be done anyway (either by using `RTC_MASK` or in hardware when `rtc_compare_set` is called).

But I guess this can be left as a field for future optimizations.

pizi-nordic commented on drivers/timer/nrf_rtc_timer.c:
```diff
@@ -266,21 +266,13 @@ u32_t _get_remaining_program_time(void)
-		rtc_prev = rtc_now;
-		rtc_now = RTC_COUNTER;
-	} while (rtc_prev != rtc_now);
+	rtc_elapsed = (RTC_COUNTER - rtc_past) & RTC_MASK;
```
Updated. Please note, that old code did the opposite (RTC_COUNTER was read before rtc_prev).

pizi-nordic commented on drivers/timer/nrf_rtc_timer.c:
```diff
@@ -371,25 +358,17 @@ void _enable_sys_clock(void)
+	elapsed = _sys_clock_tick_count;
+	compiler_barrier();
 
+	rtc_elapsed = (RTC_COUNTER - rtc_past) & RTC_MASK;
```
Updated. Similarly like in the previous case, the old code did the opposite (RTC_COUNTER was read before rtc_prev).

cvinayak commented on drivers/timer/nrf_rtc_timer.c:
```diff
@@ -266,23 +266,19 @@ u32_t _get_remaining_program_time(void)
+	rtc_elapsed = (RTC_COUNTER - rtc_past_copy) & RTC_MASK;
 
-	/*Convert number of Machine cycles to SYS TICS*/
+	/* Convert number of Machine cycles to SYS TICS */
```
SYS_TICKS. Otherwise, PR LGTM now.

pizi-nordic replied:
Fixed.

cvinayak commented:
@pizi-nordic @carlescufi LGTM

ioannisg commented on drivers/timer/nrf_rtc_timer.c:
```diff
@@ -371,25 +362,19 @@ void _enable_sys_clock(void)
+	/* Read _sys_clock_tick_count and rtc_past before RTC_COUNTER */
+	elapsed = _sys_clock_tick_count;
+	rtc_past_copy = rtc_past;
+	compiler_barrier();
```
Why do we need the compiler_barrier here? should we add an inline comment that justifies it?

pizi-nordic replied:
According to @cvinayak, the rtc_past must be read before RTC_COUNTER. Barrier ensures the correct order.

ioannisg replied:
I don't see the problem of placing the same comment in 2 locations in a file. It is really educative in this case.

pizi-nordic replied:
Ok

pizi-nordic replied:
Fixed.

ioannisg reviewed (commented):
Looks good, only few minor comments from my side.
